### PR TITLE
feat: support reporting AggregateErrors 

### PR DESCRIPTION
**Pull request by CheadleCheadle**
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions.

### Description of the Change
Addresses #4982
If an AggregateError is thrown, the base reporter will print out the AggregateError's errors. Before this change, only the Aggregate Error itself was printed. So, if the current error being printed is an Aggregate error the errors it contains will also be formatted and printed.

### Alternate Designs
I could have completely restructured the Base.list method to only loop over the failures and call a helper function that would create the format(fmt), msg, and title and then print the error, but it was much simpler to just add a conditional check to see if the current error being printed was an AggregateError and if it contained multiple errors and just 

### Why should this be in core?
Currently, nothing nice prints. When dealing with AggregateErrors it would be better if each subsequent error was printed and formatted as well.

### Benefits
The errors within the AggregateError will be nicely formatted when printed.

### Possible Drawbacks
If an AggregateError contains a lot of errors, then a lot of errors will be printed!

### Applicable issues
Nothing of note.

### CODE REVIEW

File: lib/cli/run-option-metadata.js [hunk 1]
```diff
@@ -49,7 +49,7 @@ const TYPES = (exports.types = {
     'sort',
     'watch'
   ],
-  number: ['retries', 'jobs'],
+  number: ['retries', 'jobs', 'jsonStringifyWhitespace'],
```
Comment (voxpelli) [inline lib/cli/run-option-metadata.js:52]: Minor feedback: This seems to be an unrelated addition

File: lib/reporters/base.js [hunk 2]
```diff
@@ -246,6 +246,7 @@ exports.list = function (failures) {
     if (test.err && test.err.multiple) {
       if (multipleTest !== test) {
         multipleTest = test;
+
```
Comment (voxpelli) [inline lib/reporters/base.js:249]: Preferably avoid needless styling changes like this 🙏

File: lib/reporters/base.js [hunk 3]
```diff
@@ -301,11 +302,37 @@ exports.list = function (failures) {
       }
       testTitle += str;
     });
-
```
Comment (voxpelli) [inline lib/reporters/base.js:304]: Preferably avoid needless styling changes like this 🙏

File: lib/reporters/base.js [hunk 4]
```diff
@@ -301,11 +302,37 @@ exports.list = function (failures) {
...
     Base.consoleLog(fmt, i + 1, testTitle, msg, stack);
+
+    // Handles formatting and printing Aggregate errors
+    if (test.err.errors) {
+      for (const error of test.err.errors) {
+        const testTitle = addIndent(`Error ${error.message}`, 0);
+        const stack = addIndent(error.stack, 3);
+
+        Base.consoleLog(fmt, error.message, testTitle, '', stack);
```
Comment (voxpelli) [inline lib/reporters/base.js:313]: I think the formatting of aggregated errors should use the same formatting as non-aggregated errors, so rather than formatting here locally, send `test.err.errors` to `Base.list` or such?

Helps with:

- Aggregated errors that are nested in aggregated errors
- Aggregated errors with causes in them, pending: https://github.com/mochajs/mocha/pull/4829

File: package-lock.json [hunk 5]
```diff
@@ -109,10 +109,6 @@
       },
       "engines": {
         "node": ">= 14.0.0"
-      },
-      "funding": {
-        "type": "opencollective",
-        "url": "https://opencollective.com/mochajs"
```
Comment (voxpelli) [inline package-lock.json:115]: Unrelated change, would suggest removing

File: lib/reporters/json.js [hunk 6]
```diff
@@ -153,6 +153,7 @@ function cleanCycles(obj) {
  */
 function errorJSON(err) {
   var res = {};
+
```
Comment (voxpelli) [inline lib/reporters/json.js:156]: Styling change, would suggest removing

**Review by voxpelli — CHANGES_REQUESTED** [review]
Some initial feedback

File: lib/cli/run-option-metadata.js (hunk 1 above)
Comment (CheadleCheadle) [reply lib/cli/run-option-metadata.js:52]: This relates to #5017 and involves introducing an optional --jsw flag that allows users to specify the amount of whitespace in JSON reporter output. But, I'll exclude this from the current pr to focus on addressing one feature addition.

File: lib/reporters/base.js (hunk 4 above)
Comment (CheadleCheadle) [reply lib/reporters/base.js]: Good idea! If a test contains an aggregate error it now recursively calls Base.list with those errors.

**Comment by wesleytodd** [comment]
Plus one on this. I just ran into it. Is anyone working on this anymore? It looks like maybe the feedback was addressed?

**Comment by nwalters512** [comment]
Somewhere between Node 16 and Node 20, Node's own APIs started throwing `AggregateError` in certain cases, e.g. network failures when `autoSelectFamily === true` (the default). Without proper formatting, it's very difficult to debug such errors when they occur in Mocha tests. Is there anything we can do to help this along?

**Comment by CheadleCheadle** [comment]
I addressed all the changes and sent an email to voxpelli to see if he could approve it.

**Comment by voxpelli** [comment]
I'll try to get around to it soon, else eg. @JoshuaKGoldberg can chime in?

File: lib/reporters/base.js [hunk 7]
```diff
@@ -292,17 +295,24 @@ exports.list = function (failures) {
...
-    test.titlePath().forEach(function (str, index) {
-      if (index !== 0) {
-        testTitle += '\n     ';
-      }
-      for (var i = 0; i < index; i++) {
-        testTitle += '  ';
-      }
-      testTitle += str;
-    });
+    // Incase test is an error object when recursively listing aggregate errors
```
Comment (wesleytodd) [inline lib/reporters/base.js:298]: I am not super familiar with `mocha` internals, but this comment and the line below is suspicious to me.

The comment implies this check is for when `test` is an instance of `AggregateError` but then the check is for `titlePath`. I am not saying the code is wrong, but I am saying that whatever is going on here is not clear and could probably be commented or written in a way that was more clearly addressing the issue.

**Review by wesleytodd — APPROVED** [review]

File: lib/reporters/base.js (hunk 7 above)
Comment (CheadleCheadle) [reply lib/reporters/base.js]: I went ahead and modified that if statement to check if the error is an instance of an Aggregate Error. It accomplishes the same thing but is much more clear why it is doing it.

**Comment by voxpelli** [comment]
I'm running out of time to review and merge this today, but this one together with these two are my main priority right now:

* https://github.com/mochajs/mocha/pull/5074
* #4829

File: bin/mocha.js
Comment (JoshuaKGoldberg) [inline bin/mocha.js:1]: [Question] Why the change in file permissions?

File: lib/reporters/base.js [hunk 8]
```diff
@@ -292,17 +295,24 @@ exports.list = function (failures) {
...
-    test.titlePath().forEach(function (str, index) {
-      if (index !== 0) {
-        testTitle += '\n     ';
-      }
-      for (var i = 0; i < index; i++) {
-        testTitle += '  ';
-      }
-      testTitle += str;
-    });
+    // Incase test is an AggregateError object when recursively listing errors
```
Comment (JoshuaKGoldberg) [inline lib/reporters/base.js:298]: [Docs] Nit: I think the `instanceof AggregateError` implies what this comment is trying to say. 🙂 

```suggestion
```

File: lib/reporters/base.js [hunk 9]
```diff
@@ -544,7 +554,6 @@ var objToString = Object.prototype.toString;
 function sameType(a, b) {
   return objToString.call(a) === objToString.call(b);
 }
-
```
Comment (JoshuaKGoldberg) [inline lib/reporters/base.js:547]: [Style] Nit: could you undo this unrelated change & any others please?

File: test/reporters/base.spec.js [hunk 10]
```diff
@@ -507,21 +507,34 @@ describe('Base reporter', function () {
...
 
-    it('should let you stub out console.log without effecting reporters output', function () {
-      Base.list([]);
-      baseConsoleLog.restore();
+    var errOut = stdout.join('\n').trim();
 
-      expect(baseConsoleLog, 'was called');
-      expect(console.log, 'was not called');
-    });
+    expect(errOut, 'to contain', ' Error: 1', 'Error: 2');
```
Comment (JoshuaKGoldberg) [inline test/reporters/base.spec.js:520]: [Testing] Aggregate / multiple test error reports can be tricky. I think we'd want to test more than just whether these two lines are included. Could you test more directly what the actual output is? I.e. that it includes the proper newlines, spacing, any details/messages/stacks (or not) as expected?

File: test/reporters/base.spec.js [hunk 11]
```diff
@@ -507,21 +507,34 @@ describe('Base reporter', function () {
     );
   });
 
-  describe('when reporter output immune to user test changes', function () {
-    var baseConsoleLog;
+  it('should list all the errors within an AggregateError', function () {
+    var err1 = new Error('1');
+    var err2 = new Error('2');
+    var aggErr = new AggregateError([err1, err2], '2 errors');
```
Comment (JoshuaKGoldberg) [inline test/reporters/base.spec.js:513]: [Testing] A few more test cases that should be covered...

* An `AggregateError` that only has one error
* An `AggregateError` where the constructor receives something that isn't an `Error` - which is possible per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AggregateError/AggregateError#errors

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Good start! 🙌 

I'd also like to hear from @voxpelli, but leaving some thoughts in the meantime too.

File: bin/mocha.js
Comment (CheadleCheadle) [reply bin/mocha.js:1]: Must have been accidental because I don't recall modifying bin/mocha.

File: lib/reporters/base.js (hunk 8 above)
Comment (CheadleCheadle) [reply lib/reporters/base.js:298]: Yeah, that one speaks for itself, you're right!

File: lib/reporters/base.js (hunk 9 above)
Comment (CheadleCheadle) [reply lib/reporters/base.js:547]: Will do

File: test/reporters/base.spec.js (hunk 11 above)
Comment (CheadleCheadle) [reply test/reporters/base.spec.js:513]: I went ahead and added test cases for these.

File: bin/mocha.js
Comment (JoshuaKGoldberg) [inline bin/mocha.js:1]: Still the permissions change? Is there an automatic modification happening on save or something like that?

Comment (CheadleCheadle) [reply bin/mocha.js:1]: It's possible. I'll look into it this week.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Whoops, I'd meant to post this a bit back, sorry - but looks like there are failing tests?

**Comment by JoshuaKGoldberg** [comment]
👋 Ping @CheadleCheadle, is this still something you have energy & time for?

**Comment by nwalters512** [comment]
@JoshuaKGoldberg I'm very interested in seeing this land. What's the best way for me to help, open PRs against _this_ branch? Or open a brand-new PR with these same changes?

**Comment by JoshuaKGoldberg** [comment]
Yeah, this has been stale for a while. I think it's safe to close out.

If you end up using code from this PR, please add a co-author attribution. Otherwise 🚀 free to make a new one!
